cpp/ql/src/semmle/code/cpp/controlflow/LocalScopeVariableReachability.qll

```diff
@@ -131,20 +131,27 @@ private predicate bbLoopEntryConditionAlwaysTrueAt(BasicBlock bb, int i, Control
 }
 
 /**
- * Basic block `pred` ends with a condition belonging to a loop, and that
- * condition is provably true upon entry. Basic block `succ` is a successor
- * of `pred`, and `skipsLoop` indicates whether `succ` is the false-successor
- * of `pred`.
+ * Basic block `pred` contains all or part of the condition belonging to a loop,
+ * and there is an edge from `pred` to `succ` that concludes the condition.
+ * If the edge corrseponds with the loop condition being found to be `true`, then
+ * `skipsLoop` is `false`.  Otherwise the edge corresponds with the loop condition
+ * being found to be `false` and `skipsLoop` is `true`.  Non-concluding edges
+ * within a complex loop condition are not matched by this predicate.
  */
 private predicate bbLoopConditionAlwaysTrueUponEntrySuccessor(BasicBlock pred, BasicBlock succ, boolean skipsLoop) {
-  succ = pred.getASuccessor() and
-  exists(ControlFlowNode last |
-    last = pred.getEnd() and
-    loopConditionAlwaysTrueUponEntry(_, last) and
-    if succ = pred.getAFalseSuccessor() then
-      skipsLoop = true
-    else
-      skipsLoop = false
+  exists(ControlFlowNode loop |
+    loopConditionAlwaysTrueUponEntry(loop, _) and
+    (
+      (
+        succ = loop.(Loop).getFollowingStmt() and
```

Comment:
I didn't even know we had a `getFollowingStmt`. Maybe we should deprecate it since its name is quite misleading -- it only gets the following statement in the _same_ basic block, if that exists. So I don't think this will work if the loop is the last statement in a block.

We don't have a general way to find what control-flow node(s) comes after a given AST element, but there are some ways to approximate it. This is what's done [in `FlowVar.qll`](https://github.com/Semmle/ql/blob/641126a5062112a4da4f1b6e41758adc3c1d18d7/cpp/ql/src/semmle/code/cpp/dataflow/internal/FlowVar.qll#L387-L392), although I'm not very proud of this code. When we compute the CFG in QL, all this should become less painful.

Reply:
Another issue is that in C++ there might be destructor calls following the false test in the loop condition. I think that can happen in [this case](https://github.com/Semmle/ql/blob/641126a5062112a4da4f1b6e41758adc3c1d18d7/cpp/ql/test/library-tests/qlcfg/condition_decls.cpp#L41), although I no longer remember the details.

It's probably more robust to test for the `skipsLoop = true` case by making it the negation of the `skipsLoop = false` case -- the latter should be easier to detect.

Reply:
Good point about `getFollowingStmt`.  I've changed the `skipsLoop = true` case as you suggested, added a warning on `getFollowingStmt`, and added a test where this matters.

As best I can tell the destructor in these kinds of cases occurs at the end of the loop body, not the end of the loop condition.  I've been unable to find a case where `Loop.getStmt()` doesn't do what we want, there are alternative ways I could formulate this (as in your link) but they're likely to be more complex / expensive / brittle so I'd prefer to leave it as-is unless we can find a situation where it's wrong.

Reply:
Sounds good.

Reply:
It turns out there were insufficient constraints within `bbLoopConditionAlwaysTrueUponEntrySuccessor` in my new solution, and this was at the very least causing a performance problem.  I've fixed this by looking for edges that exit the expression tree of the loop condition, which despite my concerns before is performing fine.